drag_and_drop_v2/drag_and_drop_v2.py

```diff
@@ -412,12 +414,224 @@ def default_background_image_url(self):
+                FeedbackMessage(FeedbackMessages.MISPLACED_ITEMS_RETURNED, None)
+            )
+
+        if self.attempts_remain and (misplaced_ids or missing_ids):
+            problem_feedback_message = self.data['feedback']['start']
+        else:
+            problem_feedback_message = self.data['feedback']['finish']
+
+        problem_feedback_class = self.PROBLEM_FEEDBACK_CLASSES.get(answer_correctness, None)
+        grade_feedback_class = self.GRADE_FEEDBACK_CLASSES.get(answer_correctness, None)
+
+        feedback_msgs.append(FeedbackMessage(problem_feedback_message, problem_feedback_class))
+
+        if not self.attempts_remain:
+            feedback_msgs.append(
+                FeedbackMessage(FeedbackMessages.FINAL_ATTEMPT_TPL.format(score=self.grade), grade_feedback_class)
+            )
+
+        return feedback_msgs, misplaced_ids
+
+    @staticmethod
+    def _present_overall_feedback(feedback_messages):
+        """
+        Transforms feedback messages into format expected by frontend code
+        """
+        return [
+            {"message": msg.message, "message_class": msg.message_class}
+            for msg in feedback_messages
+            if msg.message
+        ]
+
+    def _drop_item_standard(self, item_attempt):
+        """
+        Handles dropping item to a zone in standard mode.
+        """
+        item = self._get_item_definition(item_attempt['val'])
+
+        is_correct = self._is_attempt_correct(item_attempt)  # Student placed item in a correct zone
+        if is_correct:  # In standard mode state is only updated when attempt is correct
+            self.item_state[str(item['id'])] = self._make_state_from_attempt(item_attempt, is_correct)
+
+        self._mark_complete_and_publish_grade()  # must happen before _get_feedback
+        self._publish_item_dropped_event(item_attempt, is_correct)
+
+        item_feedback_key = 'correct' if is_correct else 'incorrect'
+        item_feedback = item['feedback'][item_feedback_key]
+        overall_feedback, __ = self._get_feedback()
+
+        return {
+            'correct': is_correct,
+            'finished': self._is_finished(),
+            'overall_feedback': self._present_overall_feedback(overall_feedback),
+            'feedback': item_feedback
+        }
+
+    def _drop_item_assessment(self, item_attempt):
+        """
+        Handles dropping item into a zone in assessment mode
+        """
+        if not self.attempts_remain:
+            raise JsonHandlerError(409, self.i18n_service.gettext("Max number of attempts reached"))
+
+        item = self._get_item_definition(item_attempt['val'])
+
+        is_correct = self._is_attempt_correct(item_attempt)
+        # State is always updated in assessment mode to store intermediate item positions
+        self.item_state[str(item['id'])] = self._make_state_from_attempt(item_attempt, is_correct)
+
+        self._publish_item_dropped_event(item_attempt, is_correct)
+
+        return {}
```

Comment:
@cahrens In standard mode, when item is dropped into a zone, learner get feedback (both per-item and overall) immediately.

In assessment mode, when an item is dropped into a zone, learner get no feedback at all. Instead feedback is displayed after an attempt is submitted, which is handled by `do_attempt`.